deploy-cloudgov.sh

```diff
@@ -0,0 +1,94 @@
+#!/bin/sh
+#
+# This script will attempt to create the services required
+# and then launch everything.
+#
+
+# this function will generate a random string, or bail if uuidgen is not available.
+generate_string()
+{
+  if [ -z "$1" ] ; then
+    if command -v uuidgen >/dev/null ; then
+      NEW_STRING=$(uuidgen)
+      export NEW_STRING
+    else
+      echo "cannot find uuidgen utility:  You will need to generate some random strings and put them in the CRON_KEY, HASH_SALT, and ROOT_USER_PASS environment variables, then re-run this script."
+      exit 1
+    fi
+  fi
+}
+
+# If the user does not supply required data, generate some secrets.
+generate_string "$CRON_KEY"
+CRON_KEY=${CRON_KEY:-$NEW_STRING}
+
+generate_string "$HASH_SALT"
+HASH_SALT=${HASH_SALT:-$NEW_STRING}
+
+generate_string "$ROOT_USER_PASS"
+ROOT_USER_PASS=${ROOT_USER_PASS:-$NEW_STRING}
+
+ROOT_USER_NAME=${ROOT_USER_NAME:-root}
+
+
+# function to check if a service exists
+service_exists()
+{
+  cf service "$1" >/dev/null 2>&1
+}
+
+# create services (if needed)
+if service_exists "database" ; then
+  echo database already created
+else
+  if [ "$1" = "prod" ] ; then
+    cf create-service aws-rds medium-mysql database
+  else
+    cf create-service aws-rds shared-mysql database
+  fi
+fi
+
+if service_exists "secrets" ; then
+  echo secrets already created
+else
+  yes '' | cf create-user-provided-service secrets -p "{\"CRON_KEY\": \"$CRON_KEY\", \"HASH_SALT\": \"$HASH_SALT\", \"ROOT_USER_NAME\": \"$ROOT_USER_NAME\", \"ROOT_USER_PASS\": \"$ROOT_USER_PASS\"}"
+fi
+
+if service_exists "storage" ; then
+  echo storage already created
+else
+  cf create-service s3 basic-sandbox storage
+fi
+
+# create this to figure out where to set the s3 stuff
+cf create-service-key storage storagekey
+
+# wait until the db is fully provisioned
+until cf create-service-key database test-db-ok ; do
+	echo waiting until database is live...
+	sleep 20
+done
+cf delete-service-key database test-db-ok -f
+
+# launch the apps
+cf push
+
+# make sure that the app knows where it's s3fs stuff lives
+S3INFO=$(cf service-key storage storagekey)
+S3_BUCKET=$(echo "$S3INFO" | grep '"bucket":' | sed 's/.*"bucket": "\(.*\)",/\1/')
+S3_REGION=$(echo "$S3INFO" | grep '"region":' | sed 's/.*"region": "\(.*\)",/\1/')
+cf set-env web S3_BUCKET "$S3_BUCKET"
+cf set-env web S3_REGION "$S3_REGION"
+cf delete-service-key storage storagekey -f
+# This is a bit heavyweight, but it is the only way to get the environment variables to take effect.
```

Comment:
`cf restart` should be sufficient to pick up changes in environment variables that don't influence the staging process, and it's way faster. You can also use [`cf zdt-restart`](https://docs.cloudfoundry.org/devguide/deploy-apps/rolling-deploy.html#restart) to avoid any downtime.

Reply:
Alternatively, to avoid the possibility of requests hitting the app before the env var is set and the restart has happened, the `cf push` above can include `--no-start`, and you can just do a `cf start` here.

Reply:
(Though this shouldn't be necessary if you switch to using `cf bind-service`.)

Reply:
The services are bound in the manifest.  However, the S3 stuff needs to be set as a real variable, because the stuff that uses it (apache htaccess file) has a hard time parsing json.

I did actually research the restart vs restage thing a bit, and I thought I had discovered a reason why I needed to do that, but I can no longer find whatever documentation I had found, so I will give that a whirl.

Reply:
cf zdt-restart doesn't work with multiple apps in the manifest, unfortunately.

Reply:
I guess there's no great reason why there need to be multiple apps in the manifest.  I just was rolling with that because that's how it was given to me.

Reply:
> cf zdt-restart doesn't work with multiple apps in the manifest, unfortunately.

Where did you find that info? Is it actually failing? (There's [a limitation that means you might pile up multiple app versions running at the same time](https://docs.cloudfoundry.org/devguide/deploy-apps/rolling-deploy.html#limitations), but AFAIK there's no incompatibility with multiple apps in the manifest.)